autogen/oai/client.py

```diff
@@ -52,6 +54,209 @@
 LEGACY_CACHE_DIR = ".cache"
 
 
+class Client(ABC):
+    """
+    A client class must implement the following methods:
+    - create must return a response object that implements the ClientResponseProtocol
+    - cost
+
+    This class is used to create a client that can be used by OpenAIWrapper.
+    It mimics the OpenAI class, but allows for custom clients to be used.
+    """
+
+    RESPONSE_USAGE_KEYS = ["prompt_tokens", "completion_tokens", "total_tokens", "cost", "model"]
+
+    class ClientResponseProtocol(Protocol):
+        class Choice(Protocol):
+            class Message(Protocol):
+                content: str | None
+                function_call: str | None
+
+        choices: List[Choice]
+        config_id: int
+        cost: float
+        pass_filter: bool
+        model: str
+
+    def update(self, config: Dict):
+        # update with anything here
+        pass
+
+    @abstractmethod
+    def create(self, params) -> ClientResponseProtocol:
+        pass
+
+    @abstractmethod
+    def cost(self, response: ClientResponseProtocol) -> float:
+        pass
+
+    @staticmethod
+    def get_usage(response: ClientResponseProtocol) -> Dict:
+        return None
+
+
+class OpenAIClient(Client):
+    def __init__(self, client):
+        self.client = client
+
+    def create(self, params: Dict[str, Any]) -> ChatCompletion:
+        """Create a completion for a given config using openai's client.
+
+        Args:
+            client: The openai client.
+            params: The params for the completion.
+
+        Returns:
+            The completion.
+        """
+        completions: Completions = self.client.chat.completions if "messages" in params else self.client.completions  # type: ignore [attr-defined]
```

Comment:
Is it better to use `self._client` or `self._oai_client` to distinguish the "class Client" and "OpenAI Client"? (Just for readability).

Reply:
good point, updated